test/test_sparse_csr.py

```diff
@@ -1045,6 +1045,122 @@ def test_resize(self, device, dtype):
             self.assertEqual(a.shape, (new_shape[-2], new_shape[-1]))
             self.assertEqual(a._nnz(), 5)
 
+    @skipMeta
+    @dtypes(torch.float, torch.bool)
+    @all_sparse_compressed_layouts()
+    def test_resize_as_sparse_compressed(self, device, dtype, layout):
+
+        def _check_resize_b_as_a(b, a, should_resize=True):
+            br = b.clone()
+            br.resize_as_sparse_(a)
+
+            # shape is inherited from a
+            self.assertEqual(a.shape, br.shape)
+            # other metadata is not affected
+            self.assertEqual(b.layout, br.layout)
+            self.assertEqual(b.device, br.device)
+            self.assertEqual(b.dtype, br.dtype)
+
+            def _get_compressed_plain_inds(t):
+                compressed_indices_mth, plain_indices_mth = sparse_compressed_indices_methods[t.layout]
+                return compressed_indices_mth(t), plain_indices_mth(t)
+
+            br_compressed_indices, br_plain_indices = _get_compressed_plain_inds(br)
+            br_values = br.values()
+
+            b_compressed_indices, b_plain_indices = _get_compressed_plain_inds(b)
+            a_compressed_indices, a_plain_indices = _get_compressed_plain_inds(a)
+
+            if should_resize:
+                # contents of the indices tensors should be updated with data from a, if resize is performed
+                self.assertEqual(a_plain_indices, br_plain_indices, exact_dtype=False)
+                self.assertEqual(a_compressed_indices, br_compressed_indices, exact_dtype=False)
+                # the device/dtype of indices should always be unaffected
+                self.assertEqual(b_plain_indices.dtype, br_plain_indices.dtype)
+                self.assertEqual(b_plain_indices.device, br_plain_indices.device)
+                self.assertEqual(b_compressed_indices.dtype, br_compressed_indices.dtype)
+                self.assertEqual(b_compressed_indices.device, br_compressed_indices.device)
+
+                # values are generated empty, shape is updated
+                self.assertEqual(a.values().shape, br_values.shape)
+                # the device/dtype of indices should always be unaffected
+                b_values = b.values()
+                self.assertEqual(b_values.dtype, br_values.dtype)
+                self.assertEqual(b_values.device, br_values.device)
+                # nnz will be picked up from a via new shape of values
+                self.assertEqual(a._nnz(), br._nnz())
+            else:
+                # if a resize is not done, they should remain unchanged
+                self.assertEqual(b_plain_indices, br_plain_indices, exact_device=True)
+                self.assertEqual(b_compressed_indices, br_compressed_indices, exact_device=True)
+
+                # if a resize is not triggered they should remain unchanged
+                self.assertEqual(b.values(), br_values, exact_device=True)
```

Comment:
I think we should consider `b.resize_as(b)` as an operation that guarantees the following invariants:

- `b.layout == a.layout`
- `b.shape` == `a.shape`
- `b.dense_dim() == a.dense_dim()`
- `b._nnz() == br._nnz()`
- `b.device` and `b.dtype` do not change during the `resize_as` op
- `b` is a valid sparse compressed tensor, that is, its indices satisfy the invariants of `b.layout`

but not
- `compressed_indices(b) == compressed_indices(a)`
- `plain_indices(b) == plain_indices(a)`

for efficiency and consistency. Namely, we cannot promise users that the `b` indices content after `resize_as` op is generally well-defined, for instance, currently, we have
- `b.resize_as(a).crow_indices() == orig_b.crow_indices()` iff `b.shape == a.shape`
- `b.resize_as(a).crow_indices() == a.crow_indices()` iff `b.shape != a.shape`
for instance.

When using batched and dense dimensions, these rules become even more involved. Typically, `b.resize_as(a)` follows an update of `b` indices and values, then there is no need for well-defines of `b` indices after resize, except that `b` must be a valid tensor.

So, I suggest
```suggestion
            self.assertEqual(a_plain_indices.shape, br_plain_indices.shape)
            self.assertEqual(a_compressed_indices.shape, br_compressed_indices.shape)

            # We don't check the content of br_plain_indices and br_compressed_indices 
            # because it is not well-defined (the content depends on the original
            # shape of `b` that `resize_as` ought to discard) nor needed (the
            # subsequent operation likely updates the indices and values of `b` anyway).

            # the device/dtype of indices should always be unaffected
            self.assertEqual(b_plain_indices.dtype, br_plain_indices.dtype)
            self.assertEqual(b_plain_indices.device, br_plain_indices.device)
            self.assertEqual(b_compressed_indices.dtype, br_compressed_indices.dtype)
            self.assertEqual(b_compressed_indices.device, br_compressed_indices.device)

            # values are generated empty, shape is updated
            self.assertEqual(a.values().shape, br_values.shape)
            # the device/dtype of indices should always be unaffected
            b_values = b.values()
            self.assertEqual(b_values.dtype, br_values.dtype)
            self.assertEqual(b_values.device, br_values.device)
            # nnz will be picked up from a via new shape of values
            self.assertEqual(a._nnz(), br._nnz())
```